### ISSUE TITLE
tcl-tk: Tk depends on X11

### ISSUE THREAD

**Issue by maxim-belkin** [Linuxbrew/linuxbrew#826]
I've been trying to install `tk` part of the `tcl-tk` (installation of `tcl` part proceeds smoothly).
Things I've found so far:
1. `with-x11` option uses incorrect syntax `--with-x` for `configure` command: it should be `--with-x=DIR`.
2. `patch: DATA` checks `if MacOS.version < :leopard` which should be prepended with `OS.mac? &&`
3. To find `Xlib.h` (actually, we need `X11/Xlib.h`), I used:
`rpm -ql libX11-devel | grep Xlib.h`
(system is CentOS 6.6), which returned
`/usr/include/X11/Xlib.h`
So, we need the compiler to know about `/usr/include` folder. If we append `/usr/include` to the `CFLAGS` using `ENV.append_to_cflags "-I/usr/include"` then the compiler finds an old `tcl.h` and installation fails. So, I create a link to that `X11` folder using
`system "ln", "-s", "/usr/include/X11", Dir.pwd`
as the first step of the `tk`-building phase (before the `configure` command).
4. At this point the build process fails with an error about missing `libX11.so`. In my case it is in `/usr/lib64`. So, I append this path to cflags with `ENV.append_to_cflags "-L/usr/lib64"`.
5. At this point the build process fails with an error message about unknown `libxcb.so`....
And this is where things seem weird:
```
$ ldd /usr/lib64/libX11.so
	linux-vdso.so.1 (0x00007fffd37ff000)
	libxcb.so.1 => not found
	libdl.so.2 => [linuxbrew-folder]/Cellar/glibc/2.19/lib/libdl.so.2 (0x00007f14b0c24000)
	libc.so.6 => [linuxbrew-folder]/Cellar/glibc/2.19/lib/libc.so.6 (0x00007f14b08aa000)
	[linuxbrew-folder]/Cellar/glibc/2.19/lib64/ld-linux-x86-64.so.2 (0x00007f14b0e2c000)

$ /usr/bin/ldd /usr/lib64/libX11.so
	linux-vdso.so.1 =>  (0x00007fffb9fbd000)
	libxcb.so.1 => /usr/lib64/libxcb.so.1 (0x0000003625400000)
	libdl.so.2 => /lib64/libdl.so.2 (0x0000003622c00000)
	libc.so.6 => /lib64/libc.so.6 (0x0000003622000000)
	libXau.so.6 => /usr/lib64/libXau.so.6 (0x0000003624800000)
	/lib64/ld-linux-x86-64.so.2 (0x0000003621c00000)
```
Any ideas why Linuxbrew's `ldd` does not see `libxcb.so.1`? Should I add `/usr/lib64` to `rpath`? How can I do that? Any suggestions how to proceed?

**Comment by sjackman** [comment]
Linuxbrew isn't particularly well tested with X11. My suggestion is to create a fake `xorg` keg and symlink files into that, like so:
```sh
mkdir -p `brew --cellar`/xorg/7/{include,lib}
ln -s /usr/include/X11 `brew --cellar`/xorg/7/include/
ln -s /usr/lib64/lib[xX]* `brew --cellar`/xorg/7/lib/
brew link xorg
```

> Any ideas why Linuxbrew's ldd does not see libxcb.so.1? Should I add /usr/lib64 to rpath? How can I do that? Any suggestions how to proceed?

I recommend trying to keep Linuxbrew isolated from the libraries provided by your host, to prevent version conflicts like you experience with `tcl.h`. If you want to experiment, you can add `/usr/lib64` to the dynamic library search path like so:
```sh
echo /usr/lib64 >>`brew --prefix glibc`/etc/ld.so.conf
ldconfig
```

**Comment by maxim-belkin** [comment]
> Linuxbrew isn't particularly well tested with X11.

Right, this is exactly why I've started this topic. I want to be able to compile `tk` on a Linux machine.

Fake keg + pristine `tcl-tk` formula do not work.
So, Linuxbrew's `ldd` does not see `/usr/lib64` and, therefore, does not find `libxcb` - right?
There is no `ld.so.conf` file in the `$(brew --prefix glibc)/etc/` (but there is `ld.so.cache`).
I've touched `ld.so.conf`, then:
```
$ ldconfig
$ ldconfig -v
ldconfig: Path `[linuxbrew-folder]/Cellar/glibc/2.19/lib64' given more than once
ldconfig: Can't stat [linuxbrew-folder]/Cellar/glibc/2.19/libx32: No such file or directory
[linuxbrew-folder]/Cellar/glibc/2.19/lib:
	libutil.so.1 -> libutil-2.19.so
	libnss_compat.so.2 -> libnss_compat-2.19.so
	libnss_nisplus.so.2 -> libnss_nisplus-2.19.so
	libnss_nis.so.2 -> libnss_nis-2.19.so
	libnsl.so.1 -> libnsl-2.19.so
	libnss_hesiod.so.2 -> libnss_hesiod-2.19.so
	libnss_db.so.2 -> libnss_db-2.19.so
	libnss_files.so.2 -> libnss_files-2.19.so
	libnss_dns.so.2 -> libnss_dns-2.19.so
	libanl.so.1 -> libanl-2.19.so
	libresolv.so.2 -> libresolv-2.19.so
	libcidn.so.1 -> libcidn-2.19.so
	libpcprofile.so -> libpcprofile.so
	libSegFault.so -> libSegFault.so
	librt.so.1 -> librt-2.19.so
	libpthread.so.0 -> libpthread-2.19.so
	libcrypt.so.1 -> libcrypt-2.19.so
	libmemusage.so -> libmemusage.so
	libdl.so.2 -> libdl-2.19.so
	libm.so.6 -> libm-2.19.so
	libc.so.6 -> libc-2.19.so
	ld-linux-x86-64.so.2 -> ld-2.19.so
	libthread_db.so.1 -> libthread_db-1.0.so
	libBrokenLocale.so.1 -> libBrokenLocale-2.19.so
```
I remember seeing a similar thread on `tcl-tk` but could not find it...

**Comment by sjackman** [comment]
> Any idea why /Cellar/glibc/2.19/lib64 is given more than once?

`$HOMEBREW_PREFIX/Cellar/glibc/2.19/lib64' is probably hard-coded into the search path of `ldconfig`.

Did you try adding `/usr/lib64/` to `ld.so.conf`? It's not my recommendation, as I prefer to keep the two systems separate, but you can test it for troubleshooting purposes.

> Fake keg + pristine tcl-tk formula do not work.

Fake keg is what I use. It should work. What error do you see?

**Comment by maxim-belkin** [comment]
I take it back &mdash; fake keg actually works. What happened before was that linking phase was failing while trying to link `libxml2` (I've installed `libxml2`). I assumed that it failed to link just that library, but it turned out all the libraries were not linked. So,  I did `brew unlink libxml2 && brew link xorg && brew link --overwrite libxml2` and after that I was able to build `tk`! I think you should create that fake-xorg keg that would do just those 4 lines of code. People will fix this formula for their Linux distros.
I think fake-xorg should have only the necessary stuff: `libX.so*`, `libxcb.so.1`, and so on...

**Comment by sjackman** [comment]
That's in fact not a bad idea, as a stop gap measure. I was hoping someone would package the Xorg client libraries for Linuxbrew. Until that happens, a fake-keg linking to the host libraries isn't a bad plan. It's not a priority for me right now, as I don't much use X11 with Linuxbrew. I'll tag this issue "help wanted" for now, and see if someone would like to pick it up. I'm glad to hear that the fake keg worked for you.

**Comment by maxim-belkin** [comment]
So, for true xorg libraries, someone  has to follow the steps from [here](http://www.linuxfromscratch.org/blfs/view/svn/x/x7lib.html)?

For a fake formula, this is the list of header files and libraries that were necessary for `tcl-tk` to be built on the CentOS 6.6 system that I use:


**# Symlinks in $(brew --cellar)/xorg/7/include/X11/**
```
/usr/include/X11/cursorfont.h
/usr/include/X11/keysymdef.h
/usr/include/X11/keysym.h
/usr/include/X11/Xatom.h
/usr/include/X11/Xfuncproto.h
/usr/include/X11/X.h
/usr/include/X11/Xlib.h
/usr/include/X11/Xmd.h
/usr/include/X11/Xosdefs.h
/usr/include/X11/Xproto.h
/usr/include/X11/Xprotostr.h
/usr/include/X11/Xresource.h
/usr/include/X11/Xutil.h
```
**# Symlinks in $(brew --cellar)/xorg/7/lib/**
```
/usr/lib64/libxcb.so.1
/usr/lib64/libXau.so.6
/usr/lib64/libX11.so
```

**Comment by sjackman** [comment]
> So, for true xorg libraries, someone has to follow the steps from here?

Yep, that's right.

**Comment by maxim-belkin** [comment]
Ok. Let's keep this issue open. If I find time to work on it, I'll do that.

**Comment by sjackman** [comment]
That'd be great, Maxim.

**Comment by erigler-usgs** [comment]
Kicking this to let you know others are interested.

Also, the fake keg solution allowed me to install qt (and pyqt), in addition to tcl-tk.

**Comment by sjackman** [comment]
Here's my list of linked libraries in the fake keg.
```
libX11.so               libxcb-randr.so         libxcb-xf86dri.so
libX11-xcb.so           libxcb-record.so        libxcb-xfixes.so
libXau.so               libxcb-render.so        libxcb-xinerama.so
libICE.so               libxcb-res.so           libxcb-xselinux.so
libSM.so                libxcb-screensaver.so   libxcb-xtest.so
libxcb-composite.so     libxcb-shape.so         libxcb-xvmc.so
libxcb-damage.so        libxcb-shm.so           libxcb-xv.so
libxcb-dpms.so          libxcb.so               libXext.so
libxcb-dri2.so          libxcb-sync.so          libXrender.so
libxcb-glx.so           libxcb-xevie.so         libXt.so
```

**Comment by maxim-belkin** [comment]
I don't think you need all of them. The way I determined mine was that I started with nothing linked and then ran installation gazzillion times to figure out which libraries are requested. I could debug installation and grep header files from within the installation directory, but smart ideas come last.

**Comment by maxim-belkin** [comment]
Just for myself...
```
$ ldd ~/.mozilla/plugins/libflashplayer.so
        linux-vdso.so.1 (0x00007fffbf1c6000)
        libgthread-2.0.so.0 => not found
        libX11.so.6 => not found
        libXext.so.6 => not found
        libXt.so.6 => not found
        librt.so.1 => [LINUXBREW-DIRECTORY]/Cellar/glibc/2.19/lib/librt.so.1 (0x00007f0eb8891000)
        libXcursor.so.1 => not found
        libXrender.so.1 => not found
        libssl3.so => not found
        libsmime3.so => not found
        libnss3.so => not found
        libnssutil3.so => not found
        libplds4.so => not found
        libplc4.so => not found
        libnspr4.so => not found
        libpthread.so.0 => [LINUXBREW-DIRECTORY]/Cellar/glibc/2.19/lib/libpthread.so.0 (0x00007f0eb8671000)
        libdl.so.2 => [LINUXBREW-DIRECTORY]/Cellar/glibc/2.19/lib/libdl.so.2 (0x00007f0eb846d000)
        libgtk-x11-2.0.so.0 => not found
        libgdk-x11-2.0.so.0 => not found
        libatk-1.0.so.0 => not found
        libpangoft2-1.0.so.0 => not found
        libgdk_pixbuf-2.0.so.0 => not found
        libpangocairo-1.0.so.0 => not found
        libcairo.so.2 => not found
        libpango-1.0.so.0 => not found
        libfreetype.so.6 => not found
        libfontconfig.so.1 => not found
        libgobject-2.0.so.0 => not found
        libgmodule-2.0.so.0 => not found
        libglib-2.0.so.0 => not found
        libm.so.6 => [LINUXBREW-DIRECTORY]/Cellar/glibc/2.19/lib/libm.so.6 (0x00007f0eb8176000)
        libc.so.6 => [LINUXBREW-DIRECTORY]/Cellar/glibc/2.19/lib/libc.so.6 (0x00007f0eb7dfd000)
        [LINUXBREW-DIRECTORY]/Cellar/glibc/2.19/lib64/ld-linux-x86-64.so.2 (0x00007f0eba00d000)
```

**Comment by sjackman** [comment]
Closing in favour of #892.
